src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComActivator.cs

```diff
@@ -0,0 +1,299 @@
+            [MarshalAs(UnmanagedType.BStr)] string bstrKey,
+            [MarshalAs(UnmanagedType.Interface)] out object ppvObject);
+    }
+
+    [StructLayout(LayoutKind.Sequential)]
+    public struct ComActivationContext
+    {
+        public Guid ClassId;
+        public Guid InterfaceId;
+        public string[] ActivationAssemblyList;
+    }
+
+    [StructLayout(LayoutKind.Sequential)]
+    public struct ComActivationContextInternal
+    {
+        public Guid ClassId;
+        public Guid InterfaceId;
+        public int AssemblyCount;
+        public IntPtr AssemblyList;
+        public IntPtr ClassFactoryDest;
+    }
+
+    public static class ComActivator
+    {
+        /// <summary>
+        /// Entry point for unmanaged COM activation API from managed code
+        /// </summary>
+        /// <param name="cxt">Reference to a <see cref="ComActivationContext"/> instance</param>
+        public static object GetClassFactoryForType(ComActivationContext cxt)
+        {
+            if (cxt.InterfaceId != typeof(IClassFactory).GUID
+                && cxt.InterfaceId != typeof(IClassFactory2).GUID)
+            {
+                throw new NotSupportedException();
+            }
+
+            string[] potentialAssemblies = cxt.ActivationAssemblyList ?? new string[0];
+            Type classType = FindClassType(cxt.ClassId, potentialAssemblies);
+            return new BasicClassFactory(cxt.ClassId, classType);
+        }
+
+        /// <summary>
+        /// Internal entry point for unmanaged COM activation API from native code
+        /// </summary>
+        /// <param name="cxtInt">Reference to a <see cref="ComActivationContextInternal"/> instance</param>
+        public static int GetClassFactoryForTypeInternal(ref ComActivationContextInternal cxtInt)
+        {
+            if (IsLoggingEnabled())
+            {
+                Log(
+$@"{nameof(GetClassFactoryForTypeInternal)} arguments:
+    {cxtInt.ClassId}
+    {cxtInt.InterfaceId}
+    {cxtInt.AssemblyCount}
+    0x{cxtInt.AssemblyList.ToInt64():x}
+    0x{cxtInt.ClassFactoryDest.ToInt64():x}");
+            }
+
+            try
+            {
+                string[] potentialAssembies = CreateAssemblyArray(cxtInt.AssemblyCount, cxtInt.AssemblyList);
+
+                var cxt = new ComActivationContext()
+                {
+                    ClassId = cxtInt.ClassId,
+                    InterfaceId = cxtInt.InterfaceId,
+                    ActivationAssemblyList = potentialAssembies
+                };
+
+                object cf = GetClassFactoryForType(cxt);
+                IntPtr nativeIUnknown = Marshal.GetIUnknownForObject(cf);
+                Marshal.WriteIntPtr(cxtInt.ClassFactoryDest, nativeIUnknown);
+            }
+            catch (Exception e)
+            {
+                return e.HResult;
+            }
+
+            return 0;
+        }
+
+        private static bool IsLoggingEnabled()
+        {
+#if DEBUG
+            return true;
```

Comment:
We should not be doing this logging by default to debug output in debug builds. It is annoying to keep looking at somebody else's logs in debug output that you do not care about.

Reply:
Sounds good. I think the enable/disable logic is something I need guidance on, but is really apart of the general logging/diagnostic approach. @vitek-karas any thoughts on how we would like to enable disable this?

Reply:
If it's only for development of this thing for now, then a compile time define would be enough.
If we think we need this a bit broadly... the answer depends on the discussion about logging below. If we're going to use the same infra as binder, then you'll get it for free.
If not, then you could use EventSource directly.